### ISSUE TITLE
2021/12/09 - 2021/12/17

### ISSUE THREAD

**Issue by danbi5228** [njs03332/nlp_study#13]
* 단비: 4.2.2 장까지
* 유리: 4.2, 4.4장
* 선미: 4.4 장 끝까지
* 다음주: 12/17 pm 9:00
* wiki 정리: 각자
 * 단비, 선미: python
 * 유리: chapter 3

**Comment by givitallugot** [comment]
### Squeeze, Unsqueeze

#### Squeeze
- SurnameClassifier의 forward 메소드 내에 존재
- dimension이 1인 차원을 제거 ex) [1, 20, 17] -> [20, 17]
- 해당 코드에서는 squeeze(dim=2)인데, 이는 dim=2 (즉 3번째)가 1이라면 제거하고 그렇지 않으면 제거하지 않음

#### Unsqueeze
- predict_nationality 추론 시 코드에 존재
- dimension이 1인 차원을 생성 ex) unsqueeze(0) [20, 17] -> [1, 20, 17]
- classifier에 3차원으로 학습 시켰으므로, test에서는 배치크기를 1로 생각하는 것?

### Tensor Size

```python
import torch
import torch.nn as nn

init_num_channels=77 # feature dim = voca size
num_channels=256
batch_size=128
max_surname_len=17 # surname 중 가장 긴 것

x = torch.rand(batch_size,init_num_channels,max_surname_len) # 128, 77, 17
print('input_size:', x.shape)
conv1d = nn.Conv1d(init_num_channels, num_channels, kernel_size=3) # 77, 256
print('kernel_size:', conv1d.weight.shape)
out = conv1d(x)
print('output_size:',out.shape)

# input_size: torch.Size([128, 77, 17])
# kernel_size: torch.Size([256, 77, 3])
# output_size: torch.Size([128, 256, 15])
```

- conv1d에서의 kernel size ?
(Input Size) = (Batch_size, init_num_channels, max_surname_len) = (128, 77, 17)

```python
# 배치 하나만 표현했을 때,
[ [ [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],
    ...
    [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],

                        ... ...
 
    [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],
    ...
    [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] ] ]
```

<p align="center"><img src="https://user-images.githubusercontent.com/50584633/146201988-c2505984-3e08-41f8-b1b9-450dfcb1adaa.png" alt="drawing" width="700"></p>

참고: https://sanghyu.tistory.com/24
